Add support for horizontally split draft buffer window: split-horiz

robohack opened the pull request:
Add support for horizontally split draft buffer window:  split-horiz

These changes add a new token for use with `wl-draft-buffer-style' and `wl-draft-reply-buffer-style`:  `split-horiz', which, as its name suggests, will split the message window horizontally instead of vertically like `split' does.  This style of window layout can be more convenient and space-efficient for users of wide screens.

egh commented:
What do you think about changing the behavior to use the value of `split-window-preferred-function` instead? This seems like it could be a more general solution.

egh commented:
Unrelated to this PR, you might also be interested in this from my `init.el`:

```
;; Widescreen setup
(defun egh:split-window-sensibly-widescreen (&optional window)
  "Like `split-window-sensibly', but prefer horizontal split."
  (let ((window (or window (selected-window))))
    (or (and (window-splittable-p window t)
	     ;; Split window horizontally.
	     (with-selected-window window
	       (split-window-right)))
	(and (window-splittable-p window)
	     ;; Split window vertically.
	     (with-selected-window window
	       (split-window-below)))
	(and (eq window (frame-root-window (window-frame window)))
	     (not (window-minibuffer-p window))
	     ;; If WINDOW is the only window on its frame and is not the
	     ;; minibuffer window, try to split it vertically disregarding
	     ;; the value of `split-height-threshold'.
	     (let ((split-height-threshold 0))
	       (when (window-splittable-p window)
		 (with-selected-window window
		   (split-window-below))))))))

(setq split-window-preferred-function #'egh:split-window-sensibly-widescreen)
```

robohack commented:
Generally speaking I really don't like (ab)using such general functions to support what are otherwise rather well defined and limited-scope configuration domains.  It is simply not maintainable in the long run.  I've been maintaining rather complex emacs configurations supporting multiple emacs versions and sometimes multiple subsystem versions simultaneously, for nearly 30 years now, and I have had the most trouble with such methods.  Overriding functions, and having options that call user-supplied functions for feature control is indeed more general and more flexible, but that's their downfall.  Such an approach takes what can be a simple list of very domain specific either-or flags and extends it into a far more generic API that might have to be shared with all applications running in the same emacs instance, and such APIs tend to be far less stable in my experience.

Specifically in this case Wanderlust is the only emacs application I use with one large and wide (when possible) window, er, frame, so I definitely don't want to change the setting of `split-window-preferred-function` globally, and changing it just for Wanderlust, ideally just for the one frame I use Wanderlust in, and having to write a custom function just for that use, and tying it all together on the hinge of an arguably unrelated global option and API, is the kind of complexity that makes future maintenance of personal configurations really nightmarish.

Now of course I wouldn't object to also having something like a `preferred` option that would call whatever is set in `split-window-preferred-function`, for example.  I just wouldn't likely use it.  :-)

I would object strongly to eliminating `wl-draft-reply-buffer-style` et al and relying only on `split-window-preferred-function`.

I've been using this particular patch for many years now (at least seven), and I only noticed it missing and rather critical to my general user experience with WL as I've been migrating from my old local CVS source tree to this repo and also experimenting with the MELPA release.

egh commented:
Thanks for looking into it. That seems fair. In my experience wanderlust does a few too many things itself that could be handled by general emacs behavior - e.g. splitting windows - but I suppose that is the way wanderlust works and it has made it pretty stable over many emacs releases. I use wanderlust in the same window (frame) as all my other emacs work so I don't have the same use case.

I am not a wanderlust maintainer, though, so I don't have the ability to merge anything. :)

I hope MELPA is working out.

ikazuhiro commented:
Thank you for your contribution.  I think adding new choice to `wl-draft-buffer-style` and `wl-draft-reply-buffer-style` is acceptable, but there are some points in your pull-request.

1. Your request contains modification of `wl-jump-to-draft-buffer`'s behavior.  If your `wl-jump-to-draft-buffer` is called from summary buffer, it always tries display message at point.  It seems that you treat the draft as reply in spite of your comment.  And it raises error when the point is not on summary line (i.e. the end of buffer).  I doubt it is preferable behavior especially called from draft buffer with multiple draft buffer.

2. When `wl-summary.el` is called from draft buffer, Wanderlust switches to next draft.  So, your docstring of `wl-jump-to-draft-buffer` is incorrect.  If you want to update it, please describe such behavior too.

3. Please remove comment in `wl-summary.el` and L1259 of `wl-vars.el`.

robohack commented:
I've adjusted some comments and a doc string.

The behaviour of first displaying the message at the point if called from a summary buffer is intended and strongly desired.  Unless I'm gravely mistaken it should not affect the behaviour when called from a draft buffer.

ikazuhiro commented:
> it should not affect the behaviour when called from a draft buffer

I think so neither.  But your modification to `wl-jump-to-draft-buffer` affects nehavior when called from a draft buffer.

1. Set `wl-draft-buffer-style` to `split` or `split-horiz`.
2. go to summary buffer.
3. Run `wl-jump-to-draft-buffer` twice.  draft buffer is too splitted.

I think whole your modification to `wl-jump-to-draft-buffer` should be encapsulated by `(when (eq major-mode 'wl-summary-mode)...)`

robohack commented:
Ah, now I see what you mean!

ikazuhiro commented:
1. Please revive `(switch-to-buffer buf)` for the case called from buffer other than draft mode.
2. What is purpose of changing `wl-summary-write`?

ikazuhiro commented:
(supplement for [the previous comment](https://github.com/wanderlust/wanderlust/pull/153#issuecomment-432863175))

`wl-summary-write` is a command to write a new draft message, not a reply.  So, I think there is no need to display a message at the point.  If you need to display the "parent" message, you should reply to it (by calling wl-summary-reply).

robohack commented:
Your prompting has convinced me to take a deeper look at my old patches and I decided to re-write them from scratch.

In the process I discovered an old bug that had been confusing the approach I was using to make my specific use-case work as I expected it too.  I've also cleaned up behaviour for calls to create or return to draft buffers from the folder window, depending on what other windows might be visible.

I believe the current state of this branch now describes more clearly what I've been trying to achieve, and my testing so far suggests that it hasn't broken any other mode that others might use.

One remaining refinement might be to use `wl-message-window-size` where I've indicated.

ikazuhiro commented:
1. In  `wl-draft-do-reply-buffer-window-styling` and `wl-draft-do-buffer-window-styling`:
Plase handle the case that `(wl-folder-jump-to-previous-summary)` does not jump to a summary buffer (e.g. the case that there is no Summary buffer).
2. When `wl-draft-reply-buffer-style` is `msg-split` or `msg-split-horiz` and you type `w` (`wl-summary-wrte`) or `v w` (`wl-summary-toggle-disp-msg`, `wl-summary-wrte`) the result is different from each other.  Is it intended?
3. In `wl-draft-buffer-style` and `wl-draft-reply-buffer-style`:
Docstring of possible values seems exchanged.
4. In `wl-draft-buffer-style`:
Revert changing of the default value.  If you really think the default value should be changed for most users, create another pull-request.
5. In `wl-summary-write`:
Revert the change.  If you really think it should be applied for most users, create another pull-request.

robohack commented:
Thank you for your review!

1. I'm not entirely sure I've done this correctly, but I think I've now made it consistent.

2. yes, this is what I intended, at least for msg-split-horiz (when writing a new message, and if the message buffer is visible, split the message buffer horizontally, else split the whole Summary window horizontally).  So it is effectively the same as `split-horiz` if the message buffer is not visible.  Personally I don't like the way `split` worked (before it was broken, and now after I've fixed it to work the way it originally worked).  I've only invented `split-horiz` to mirror its bizarre (to me) behaviour.  I would rather `split` and `split-horiz` be replaced with their new `msg-split` and `msg-split-horiz` variants, but for now they serve to illustrate the issue.

3.  fixed, thanks!

4. oops.  fixed.

5. hmmm...  I'm getting rusty -- I forgot how `interactive-p` worked!  fixed.


BTW, I'm trying to learn to use `(when cond expr1 expr2 ...)` instead of `(if cond (progn expr1 expr2 ...))` in Emacs lisp.  I'm not sure if that's consistently done in Wanderlust code or not.  It still looks odd to me to have a `(when ...)` followed by an `(if ...)` and I'm not sure I wouldn't prefer just to avoid `when` (it is, after all, only just over 20 years since it entered mainstream Emacs subr.el!).

ikazuhiro commented:
About 3. in [the previous comment](https://github.com/wanderlust/wanderlust/pull/153#issuecomment-436644971), I may have misread your code.  In the current your pull-request, definitions of `wl-draft-reply-buffer-style` and `wl-draft-buffer-style` are inconsistent with `wl-draft-do-reply-buffer-window-styling` and `wl-draft-do-buffer-window-styling`.  Probably exchanged content was not docstring, but candidate values.
After the fix of the above, if you can, please squash commits into one.

robohack commented:
I think I have settings, names, and docstrings more consistent now!

Maybe it is better to squash the commits as part of the merge?  (set "merge button" option in the repository settings to "Allow squash merging")

[https://blog.github.com/2016-04-01-squash-your-commits/](url)

ikazuhiro commented:
I didn't know this GitHub's feature and your pull-request is merged with additional minor fix.
thank you!

robohack commented:
Thank you!
